### PR TITLE
Add allow-lists for headers and query parameters

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
LogOptions now contains allow-lists for headers and query parameters.
Any header and/or query parameter not in the allow lists, combined with
the default set, will have its value written as REDACTED to the log.

- [ ] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [ ] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [ ] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
Fixes https://github.com/Azure/azure-sdk-for-go/issues/15330

File: sdk/azcore/runtime/policy_logging.go [hunk 1]
```diff
@@ -30,7 +33,49 @@ func NewLogPolicy(o *policy.LogOptions) policy.Policy {
 	if o == nil {
 		o = &policy.LogOptions{}
 	}
-	return &logPolicy{options: *o}
+	// construct default hash set of allowed headers
+	allowedHeaders := map[string]struct{}{
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/policy_logging.go:37]: This set was taken from the [.NET](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/DiagnosticsOptions.cs#L21) implementation.

File: sdk/azcore/runtime/policy_logging.go [hunk 2]
```diff
@@ -30,7 +33,49 @@ func NewLogPolicy(o *policy.LogOptions) policy.Policy {
...
+		"x-ms-client-request-id":        {},
+		"x-ms-return-client-request-id": {},
+	}
+	// add any caller-specified allowed headers to the set
+	for _, ah := range o.AllowedHeaders {
+		allowedHeaders[strings.ToLower(ah)] = struct{}{}
+	}
+	// now do the same thing for query params
+	allowedQP := map[string]struct{}{
+		"api-version": {},
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/policy_logging.go:69]: I added this as I think it's valuable (.NET doesn't specify any).  Let me know if there are other query params we know are safe to add.

File: sdk/azcore/runtime/response.go [hunk 3]
```diff
@@ -146,42 +143,3 @@ func DecodeByteArray(s string, v *[]byte, format Base64Encoding) error {
 		return fmt.Errorf("unrecognized byte array format: %d", format)
 	}
 }
-
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/response.go:149]: This content was moved to policy_logging.go

**Review by JeffreyRichter — APPROVED** [review]
This looks great. One potential issue...
The customer create a LogOptions and sets some headers/QPs. This then get passed to the client ctor which wants to add more headers/QPs - this is a problem because the ctor should NOT modify the customer's Options instance. The ctor would have to make a deep copy of itself, then add the service-specific headers/QPs and then pass the copy down to NewPipeline.

We had a similar problem with customer & client lib policies and we made NewPipeline take 2 extra []Policy parameters. 

Maybe we need a client-only struct with extra fields in it (for future proofing) that customers don't know about but client lib code can use to augment the customer's client options without mutating it in any way.

File: sdk/azcore/policy/policy.go [hunk 4]
```diff
@@ -56,6 +56,14 @@ type LogOptions struct {
 	// The default value is false.
 	// NOTE: enabling this can lead to disclosure of sensitive information, use with care.
 	IncludeBody bool
+
+	// AllowedHeaders is the slice of headers to log with their values intact.
+	// All headers not in the slice will have their values REDACTED.
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply sdk/azcore/policy/policy.go:61]: Fixed.

**Comment by jhendrixMSFT** [comment]
This implementation doesn't mutate the customer-provided slices.

**Review by chlowell — APPROVED** [review]

**Pull request merged**